<a href="https://colab.research.google.com/github/lacey-conrad/WGCNA-work/blob/main/Gene_coexp_GCNN.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt 
import seaborn as sns

import tensorflow as tf


from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn import metrics
from sklearn.metrics import confusion_matrix, accuracy_score
from sklearn.preprocessing import LabelEncoder

from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense
from tensorflow import keras
from tensorflow.keras.layers import LeakyReLU

# 1. Reading in the dataset:

In [2]:
#df = pd.read_table('thermo_capacity_counts.txt')

In [3]:
df_gastroc = pd.read_csv('gastroc_raw_counts.csv')

In [4]:
#df_copy = df.copy()

In [5]:
df_gastroc.head()

Unnamed: 0  LN.F1.113.2be  LN.F1.113.aa  LN.F1.113.ae  LN.F1.113.ag  \
0  LOC121821266              0             0             0             0   
1  LOC102911281              0             0             0             0   
2  LOC102926907              0             0             0             0   
3  LOC121821617              0             0             0             1   
4  LOC102925994              3             4             3             4   

   LN.F1.113.ak  LN.F1.113.am  LN.F1.113.b  LN.F1.113.h  LN.F1.113.p  ...  \
0             0             0            0            0            0  ...   
1             0             0            0            0            0  ...   
2             0             0            0            0            0  ...   
3             0             0            0            0            0  ...   
4             7             7            9            2            6  ...   

   ME.F1.145.ad  ME.F1.145.d  ME.F1.145.g  ME.F1.145.i  ME.F1.145.w  \
0             0            0            0            0            0   
1             0            0            0            0            0   
2             0            0            0            0            0   
3             0            1            0            1            0   
4             3            9            5            6           11   

   ME.F1.146.aa  ME.F1.146.ab  ME.F1.146.ac  ME.F1.146.be  ME.F1.146.bf  
0             0             0             0             0             0  
1             0             0             0             0             0  
2             0             0             0             0             0  
3             0             0             1             1             0  
4             2             2             4             2             5  

[5 rows x 48 columns]

In [6]:
df_gastroc['Row_Mean'] = df_gastroc.mean(axis=1)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  """Entry point for launching an IPython kernel.


In [7]:
df_gastroc.head()

Unnamed: 0  LN.F1.113.2be  LN.F1.113.aa  LN.F1.113.ae  LN.F1.113.ag  \
0  LOC121821266              0             0             0             0   
1  LOC102911281              0             0             0             0   
2  LOC102926907              0             0             0             0   
3  LOC121821617              0             0             0             1   
4  LOC102925994              3             4             3             4   

   LN.F1.113.ak  LN.F1.113.am  LN.F1.113.b  LN.F1.113.h  LN.F1.113.p  ...  \
0             0             0            0            0            0  ...   
1             0             0            0            0            0  ...   
2             0             0            0            0            0  ...   
3             0             0            0            0            0  ...   
4             7             7            9            2            6  ...   

   ME.F1.145.d  ME.F1.145.g  ME.F1.145.i  ME.F1.145.w  ME.F1.146.aa  \
0            0            0            0            0             0   
1            0            0            0            0             0   
2            0            0            0            0             0   
3            1            0            1            0             0   
4            9            5            6           11             2   

   ME.F1.146.ab  ME.F1.146.ac  ME.F1.146.be  ME.F1.146.bf  Row_Mean  
0             0             0             0             0  0.000000  
1             0             0             0             0  0.000000  
2             0             0             0             0  0.000000  
3             0             1             1             0  0.404255  
4             2             4             2             5  4.319149  

[5 rows x 49 columns]

In [8]:
df_gastroc = df_gastroc[df_gastroc.Row_Mean >= 10]

In [9]:
df_gastroc.head()

Unnamed: 0  LN.F1.113.2be  LN.F1.113.aa  LN.F1.113.ae  LN.F1.113.ag  \
32  LOC102903073             10            12             5             6   
42  LOC102904318             27            35            21            25   
48       Ppp2r1a            336           264           175           298   
55  LOC102906476            384           340           407           446   
60  LOC102922856              6            13            21            21   

    LN.F1.113.ak  LN.F1.113.am  LN.F1.113.b  LN.F1.113.h  LN.F1.113.p  ...  \
32             7             8            9           12           11  ...   
42            25            24           37           23           25  ...   
48           425           288          230          326          262  ...   
55           526           577          343          500          447  ...   
60            17            15           18           16           18  ...   

    ME.F1.145.d  ME.F1.145.g  ME.F1.145.i  ME.F1.145.w  ME.F1.146.aa  \
32           14           11           23           20             6   
42           20           27           37           35            18   
48          182          188          426          311           290   
55          364          332          642          588           350   
60            8           18           25           32            16   

    ME.F1.146.ab  ME.F1.146.ac  ME.F1.146.be  ME.F1.146.bf    Row_Mean  
32             8             6            10            19   10.680851  
42            18            19            13            22   22.787234  
48           324           266           204           306  292.148936  
55           371           331           345           461  400.617021  
60             8            11            14            24   16.468085  

[5 rows x 49 columns]

In [10]:
# Keep!!!
df_gastroc.drop("Row_Mean", axis=1, inplace=True)

In [11]:
df_gastroc = df_gastroc.T

In [12]:
df_gastroc.columns = df_gastroc.iloc[0]

In [13]:
df_gastroc = df_gastroc.iloc[1: , :]

In [14]:
df_gastroc.head()

Unnamed: 0    LOC102903073 LOC102904318 Ppp2r1a LOC102906476 LOC102922856  \
LN.F1.113.2be           10           27     336          384            6   
LN.F1.113.aa            12           35     264          340           13   
LN.F1.113.ae             5           21     175          407           21   
LN.F1.113.ag             6           25     298          446           21   
LN.F1.113.ak             7           25     425          526           17   

Unnamed: 0    LOC102912116 Ppp4r2 Myadm LOC107402623 Cacng6  ... EI127_mgp04  \
LN.F1.113.2be           13     33    83           23    111  ...       11886   
LN.F1.113.aa             7     51    36           39    184  ...       12638   
LN.F1.113.ae            10     43    43           30    200  ...        9748   
LN.F1.113.ag            16     54    61           23    131  ...       16001   
LN.F1.113.ak            19     58   132           51    259  ...       13241   

Unnamed: 0    EI127_mgt17 EI127_mgt18 EI127_mgt19 EI127_mgp03 EI127_mgp02  \
LN.F1.113.2be         276         122          92        6169        5420   
LN.F1.113.aa          371         207          81        7024        6495   
LN.F1.113.ae          272         135          67        5841        5400   
LN.F1.113.ag          495         230         116        9873        7522   
LN.F1.113.ak          337         180          91        7773        6657   

Unnamed: 0    EI127_mgt20 EI127_mgp01 EI127_mgt21 EI127_mgt22  
LN.F1.113.2be        2144       16304         173        1511  
LN.F1.113.aa         1673       16561         142        1280  
LN.F1.113.ae         1414       13018         103         955  
LN.F1.113.ag         2011       21779         196        1599  
LN.F1.113.ak         1650       16713         109        1130  

[5 rows x 9423 columns]

In [15]:
df_gastroc = df_gastroc.reset_index()

In [16]:
df_gastroc.head()

Unnamed: 0          index LOC102903073 LOC102904318 Ppp2r1a LOC102906476  \
0           LN.F1.113.2be           10           27     336          384   
1            LN.F1.113.aa           12           35     264          340   
2            LN.F1.113.ae            5           21     175          407   
3            LN.F1.113.ag            6           25     298          446   
4            LN.F1.113.ak            7           25     425          526   

Unnamed: 0 LOC102922856 LOC102912116 Ppp4r2 Myadm LOC107402623  ...  \
0                     6           13     33    83           23  ...   
1                    13            7     51    36           39  ...   
2                    21           10     43    43           30  ...   
3                    21           16     54    61           23  ...   
4                    17           19     58   132           51  ...   

Unnamed: 0 EI127_mgp04 EI127_mgt17 EI127_mgt18 EI127_mgt19 EI127_mgp03  \
0                11886         276         122          92        6169   
1                12638         371         207          81        7024   
2                 9748         272         135          67        5841   
3                16001         495         230         116        9873   
4                13241         337         180          91        7773   

Unnamed: 0 EI127_mgp02 EI127_mgt20 EI127_mgp01 EI127_mgt21 EI127_mgt22  
0                 5420        2144       16304         173        1511  
1                 6495        1673       16561         142        1280  
2                 5400        1414       13018         103         955  
3                 7522        2011       21779         196        1599  
4                 6657        1650       16713         109        1130  

[5 rows x 9424 columns]

In [17]:
df_gastroc = df_gastroc.rename(columns={'index': 'Treatment'})

In [18]:
df_gastroc.head()

Unnamed: 0      Treatment LOC102903073 LOC102904318 Ppp2r1a LOC102906476  \
0           LN.F1.113.2be           10           27     336          384   
1            LN.F1.113.aa           12           35     264          340   
2            LN.F1.113.ae            5           21     175          407   
3            LN.F1.113.ag            6           25     298          446   
4            LN.F1.113.ak            7           25     425          526   

Unnamed: 0 LOC102922856 LOC102912116 Ppp4r2 Myadm LOC107402623  ...  \
0                     6           13     33    83           23  ...   
1                    13            7     51    36           39  ...   
2                    21           10     43    43           30  ...   
3                    21           16     54    61           23  ...   
4                    17           19     58   132           51  ...   

Unnamed: 0 EI127_mgp04 EI127_mgt17 EI127_mgt18 EI127_mgt19 EI127_mgp03  \
0                11886         276         122          92        6169   
1                12638         371         207          81        7024   
2                 9748         272         135          67        5841   
3                16001         495         230         116        9873   
4                13241         337         180          91        7773   

Unnamed: 0 EI127_mgp02 EI127_mgt20 EI127_mgp01 EI127_mgt21 EI127_mgt22  
0                 5420        2144       16304         173        1511  
1                 6495        1673       16561         142        1280  
2                 5400        1414       13018         103         955  
3                 7522        2011       21779         196        1599  
4                 6657        1650       16713         109        1130  

[5 rows x 9424 columns]

In [19]:
df_gastroc['Treatment'] = df_gastroc['Treatment'].str.split('.').str[0]

In [20]:
df_gastroc.head()

Unnamed: 0 Treatment LOC102903073 LOC102904318 Ppp2r1a LOC102906476  \
0                 LN           10           27     336          384   
1                 LN           12           35     264          340   
2                 LN            5           21     175          407   
3                 LN            6           25     298          446   
4                 LN            7           25     425          526   

Unnamed: 0 LOC102922856 LOC102912116 Ppp4r2 Myadm LOC107402623  ...  \
0                     6           13     33    83           23  ...   
1                    13            7     51    36           39  ...   
2                    21           10     43    43           30  ...   
3                    21           16     54    61           23  ...   
4                    17           19     58   132           51  ...   

Unnamed: 0 EI127_mgp04 EI127_mgt17 EI127_mgt18 EI127_mgt19 EI127_mgp03  \
0                11886         276         122          92        6169   
1                12638         371         207          81        7024   
2                 9748         272         135          67        5841   
3                16001         495         230         116        9873   
4                13241         337         180          91        7773   

Unnamed: 0 EI127_mgp02 EI127_mgt20 EI127_mgp01 EI127_mgt21 EI127_mgt22  
0                 5420        2144       16304         173        1511  
1                 6495        1673       16561         142        1280  
2                 5400        1414       13018         103         955  
3                 7522        2011       21779         196        1599  
4                 6657        1650       16713         109        1130  

[5 rows x 9424 columns]

In [21]:
cleanup_nums = {"Treatment": {"LN": 0, "ME": 1}}

In [22]:
df_gastroc = df_gastroc.replace(cleanup_nums)
df_gastroc.head()

Unnamed: 0  Treatment LOC102903073 LOC102904318 Ppp2r1a LOC102906476  \
0                   0           10           27     336          384   
1                   0           12           35     264          340   
2                   0            5           21     175          407   
3                   0            6           25     298          446   
4                   0            7           25     425          526   

Unnamed: 0 LOC102922856 LOC102912116 Ppp4r2 Myadm LOC107402623  ...  \
0                     6           13     33    83           23  ...   
1                    13            7     51    36           39  ...   
2                    21           10     43    43           30  ...   
3                    21           16     54    61           23  ...   
4                    17           19     58   132           51  ...   

Unnamed: 0 EI127_mgp04 EI127_mgt17 EI127_mgt18 EI127_mgt19 EI127_mgp03  \
0                11886         276         122          92        6169   
1                12638         371         207          81        7024   
2                 9748         272         135          67        5841   
3                16001         495         230         116        9873   
4                13241         337         180          91        7773   

Unnamed: 0 EI127_mgp02 EI127_mgt20 EI127_mgp01 EI127_mgt21 EI127_mgt22  
0                 5420        2144       16304         173        1511  
1                 6495        1673       16561         142        1280  
2                 5400        1414       13018         103         955  
3                 7522        2011       21779         196        1599  
4                 6657        1650       16713         109        1130  

[5 rows x 9424 columns]

In [23]:
####################################
# Keep all code above!!! #######
###################################

In [24]:
X = df_gastroc.drop(labels=['Treatment'], axis = 1)
y = df_gastroc.Treatment

In [25]:
y = y.astype(int)

In [26]:
#genes = list(X.columns)

In [27]:
#genes

In [28]:
##scaler = StandardScaler()
##scaler.fit(X)
#X = scaler.transform(X)

In [29]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state = 42)

In [30]:
X_train  = X_train.astype(int)
X_test  = X_test.astype(int)

In [31]:
model = tf.keras.models.Sequential()

Hyperparameters:

number of neurons: 128

activation function: ReLU

input_shape: (9423, )

In [32]:
model.add(tf.keras.layers.Dense(units = 128, activation = 'relu', input_shape = (9423,)))


Add a dropout layer

In [33]:
model.add(tf.keras.layers.Dropout(0.2))

Add an output layer

In [34]:
model.add(tf.keras.layers.Dense(units = 10, activation = 'softmax'))

Model compiler

In [35]:
model.compile(optimizer = 'adam', loss = 'sparse_categorical_crossentropy', metrics = ['sparse_categorical_accuracy'])


In [36]:
#model.fit(X_train, y_train, epochs =5)

Training the ANN

In [37]:
# Training the ANN:
model.fit(X_train, y_train, epochs = 10, batch_size = 10, verbose=1)

Epoch 1/10
4/4 [==============================] - 1s 21ms/step - loss: 5234.8140 - sparse_categorical_accuracy: 0.2973
Epoch 2/10
4/4 [==============================] - 0s 21ms/step - loss: 974.8487 - sparse_categorical_accuracy: 0.5946
Epoch 3/10
4/4 [==============================] - 0s 15ms/step - loss: 412.5680 - sparse_categorical_accuracy: 0.7838
Epoch 4/10
4/4 [==============================] - 0s 18ms/step - loss: 1323.9080 - sparse_categorical_accuracy: 0.6486
Epoch 5/10
4/4 [==============================] - 0s 22ms/step - loss: 1758.4828 - sparse_categorical_accuracy: 0.6216
Epoch 6/10
4/4 [==============================] - 0s 13ms/step - loss: 999.4434 - sparse_categorical_accuracy: 0.6757
Epoch 7/10
4/4 [==============================] - 0s 11ms/step - loss: 608.6284 - sparse_categorical_accuracy: 0.6757
Epoch 8/10
4/4 [==============================] - 0s 15ms/step - loss: 1262.2539 - sparse_categorical_accuracy: 0.5405
Epoch 9/10
4/4 [==============================] - 0s

In [39]:
test_loss, test_accuracy = model.evaluate(X_test, y_test)

1/1 [==============================] - 0s 287ms/step - loss: 192.8473 - sparse_categorical_accuracy: 0.8000


In [40]:
test_loss

192.84725952148438

In [41]:
test_accuracy

0.800000011920929

In [ ]:
model = Sequential()
model.add(Dense(50, activation='relu'))
model.add(Dense(500, activation='relu'))
model.add(Dense(1, activation = 'sigmoid'))
model.compile(optimizer='adam', loss = 'binary_crossentropy', metrics=['accuracy'])
model.fit(X_train, y_train.to_numpy(), epochs = 10, batch_size = 10, verbose=1) 

In [ ]:
y_pred = model.predict_classes(X_test)
model.evaluate(X_test, y_test.to_numpy())

In [ ]:
y_pred

In [ ]:
confusion_matrix(y_test, y_pred)

In [ ]:
accuracy_score(y_test, y_pred)

In [ ]:
# Model#3 - larger batch size
# Input layer: 32 nodes
# Hidden layer 1: 64 nodes
# Optimizer: Adam
# Loss: Binary Crossentropy
# Metrics: Accuracy
# Epochs: 10
# Batch size: 500

In [ ]:
model = Sequential()
model.add(Dense(32, activation='relu'))
model.add(Dense(64, activation='relu'))
model.add(Dense(1, activation = 'sigmoid'))
model.compile(optimizer='adam', loss = 'binary_crossentropy', metrics=['accuracy'])
model.fit(X_train, y_train.to_numpy(), epochs = 10, batch_size = 500, verbose=1) 

In [ ]:
y_pred = model.predict_classes(X_test)
model.evaluate(X_test, y_test.to_numpy())

In [ ]:
y_pred

In [ ]:
confusion_matrix(y_test, y_pred)

In [ ]:
accuracy_score(y_test, y_pred)

In [ ]:
# Model#4 - More epochs
# Input layer: 32 nodes
# Hidden layer 1: 64 nodes
# Optimizer: Adam
# Loss: Binary Crossentropy
# Metrics: Accuracy
# Epochs: 100
# Batch size: 10

In [ ]:
model = Sequential()
model.add(Dense(32, activation='relu'))
model.add(Dense(64, activation='relu'))
model.add(Dense(1, activation = 'sigmoid'))
model.compile(optimizer='adam', loss = 'binary_crossentropy', metrics=['accuracy'])
model.fit(X_train, y_train.to_numpy(), epochs = 100, batch_size = 10, verbose=1) 

In [ ]:
y_pred = model.predict_classes(X_test)
model.evaluate(X_test, y_test.to_numpy())

In [ ]:
y_pred

In [ ]:
confusion_matrix(y_test, y_pred)

In [ ]:
accuracy_score(y_test, y_pred)

# Insights derived from altering model hyperparameters:

- A smaller batch size produced more accurate results, but took longer to run.

- Adding more nodes in the hidden layers did not seem to improve the accuracy of the model.

- Increasing the number of epochs decreased the accuracy.

- My best model was number one:
 
 Model#1
 
 Input layer: 32 nodes
 
 Hidden layer 1: 64 nodes
 
 Optimizer: Adam
 
 Loss: Binary Crossentropy
 
 Metrics: Accuracy
 
 Epochs: 10
 
 Batch size: 10
 
True negatives: 1514 (Those who my model predicted would not exit, who did not actually exit)

False negatives: 79 (Those who my model predicted would exit, but who did not actually exit)

False positive: 200 (Those who my model predicted would not exit, but who did actually exit)

True positives: 207 (Those who my model predicted would exit the bank, who did actually exit the bank)

 accuracy of model: 86%